### ISSUE TITLE
Exception to ES.63 for base type without any data member

### ISSUE THREAD

**Issue by hwisungi** [isocpp/CppCoreGuidelines#2206]
After fixing a false negative bug in msvc code analysis, some library code were newly tagged for violations of ES.63 rule. One of the patterns that received those warnings was brought to our attention as potential candidate of exception to the rule. Here is the pattern:

```
    template<typename T>
    struct Tag { /* No data member */ };

    template<typename T>
    constexpr int value(Tag<T>);

    template<typename T>
    struct S : Tag<T> {
        int x;
    };

    template<typename T>
    void f() {
        S<T> s;
        value(s);
    }
```

Serving as a tag type, Tag does not need any data from the source object to perform its operations. Thus, it should be more efficient for a new instance of T to be created from the source object than creating a pointer / reference to the source object.

So, it was suggested to add an exception for this pattern to rule ES.63.

**Comment by GabrielDosReis** [comment]
Yes, there is not much cause for slicing concerns in these scenarios of tag-based selection.

Suggestion: warn only if both the base class and the derive class have non-static data members.

**Comment by prathameshatkare** [comment]
suggested to consider an exception to this rule for tag types because:

Tag Types (like Tag<T>) don't have data members: They serve purely as type markers and carry no data. Hence, copying or passing them by value is neither expensive nor inefficient.

Copying is lightweight: Since Tag<T> (and other similar tag types) don't contain any member data, the cost of copying them is essentially zero. There's no advantage in using a pointer or reference over passing by value in this case.

Semantics and design: Tag types are often used as simple identifiers for types or traits, and it's semantically more straightforward to pass them by value. Creating pointers or references to tag types introduces unnecessary complexity without any benefit in performance.

**Comment by BjarneStroustrup** [comment]
In the original example, what useful action could value() perform? Anything more than "not implemented"?

**Comment by hsutter** [comment]
Editors call: Thanks! We agree, fixed.

**Comment by eyalroz** [comment]
@hwisungi , @prathameshatkare , @GabrielDosReis  : please note my stronger objection regarding this rule: #2336 . Your comments there are welcome, especially regarding my claim that it is not possible to impugn slicing of an arbitrary class without making strong assumptions about the class being sliced.

**Comment by jwakely** [comment]
Presumably referring to #2336

**Comment by eyalroz** [comment]
@jwakely : Yes, sorry about that.
